### PR TITLE
Moving cooperative group validation

### PR DESCRIPTION
**Pull request by mmigdal-nv**
- Increasing kernel dynamic smem limit earlier (compilation/recompilation)
- Updating the stored dynamic smem limit to avoid redundant calls
- `validateCooperativeLaunch` needs to be done after the limit increase, otherwise it can lead to false positives

### CODE REVIEW

**Comment by mmigdal-nv** [comment]
!build

**Comment by mmigdal-nv** [comment]
Got the launch error from CUDA when testing https://github.com/NVIDIA/Fuser/pull/226 (but it's not related to this MR specifically)

File: csrc/executor.cpp [hunk 1]
```diff
@@ -1513,6 +1508,13 @@ std::vector<at::Tensor> FusionExecutor::runFusion(
           nullptr));
     } else {
       FUSER_PERF_SCOPE("ExecutorRunFusion::cuLaunchCooperativeKernel");
+
+      // Validation must happen after setting the attribute
+      // CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES, otherwise we can get
+      // failures.
+      validateCooperativeLaunch(
```
Comment (naoyam) [inline csrc/executor.cpp:1515]: Hmm, I wonder if this is helpful if needs to be done here. The user would just always see some error, either from us or the CUDA runtime. If there's no error, the check is effectively done twice redundantly. If this is done before, and only once for each compiled kernel, it was useful as we could report the error early and the check is just one time cost.

Is it possible to do the check with the maximum shmem size when necessary?

Comment (mmigdal-nv) [reply csrc/executor.cpp]: @naoyam I updated the PR (and description) to fail earlier - and do less checks

Comment (mmigdal-nv) [reply csrc/executor.cpp]: I am find dropping the PR though

**Comment by naoyam** [comment]
@mmigdal-nv Thanks for the fix. I further cleaned up the related code.

File: csrc/executor.h [hunk 2]
```diff
@@ -106,7 +106,7 @@ class TORCH_CUDA_CU_API FusionExecutor : public NonCopyable {
   // function to query whether a `FusionExecutor` has a compiled kernel to
   // execute
   bool compiled() const {
-    return fusion_id_ != -1 && lowered_;
```
Comment (naoyam) [inline csrc/executor.h:109]: Having `lowered_` doesn't mean the compilation is completed

File: csrc/executor.h [hunk 3]
```diff
@@ -254,13 +254,13 @@ class TORCH_CUDA_CU_API FusionExecutor : public NonCopyable {
   LaunchParams computeLaunchParams(
       const LaunchParams& launch_constraints,
       ExpressionEvaluator& expr_eval,
-      const int warp_size);
```
Comment (naoyam) [inline csrc/executor.h:257]: `int` to `int64_t` replacement

File: csrc/executor.h [hunk 4]
```diff
@@ -311,34 +311,50 @@ class TORCH_CUDA_CU_API FusionExecutor : public NonCopyable {
...
+
+  //! Check if the shared memory size can be expandable to accommodate
+  //! the given dynamic size. The total shared memory size consumed
+  //! would be the sum of the static and dynamic sizes.
+  void validateDynamicSmemSize(int64_t dynamic_smem_size);
+
+  //! Make sure the dynamic shared memory size is at least as large as
+  //! the given size
+  int64_t ensureAvailableDynamicSmemSize(int64_t dynamic_smem_size);
+
```
Comment (naoyam) [inline csrc/executor.h:328]: These are the main additions

File: csrc/executor.h [hunk 5]
```diff
@@ -311,34 +311,50 @@ class TORCH_CUDA_CU_API FusionExecutor : public NonCopyable {
...
  private:
   CompileOptions options_;
 
-  //! Current configured total shared mem size from cudaDeviceProp
-  size_t configured_device_smem_ = std::numeric_limits<size_t>().max();
```
Comment (naoyam) [inline csrc/executor.h:318]: unused

File: csrc/executor_utils.h [hunk 6]
```diff
@@ -28,39 +28,41 @@
...
 
-#define NVRTC_SAFE_CALL(x)                               \
-  do {                                                   \
-    nvrtcResult _result = x;                             \
-    if (_result != NVRTC_SUCCESS) {                      \
-      std::cerr << "\nerror: " #x " failed with error "  \
-                << nvrtcGetErrorString(_result) << '\n'; \
-      exit(1);                                           \
-    }                                                    \
+#define NVRTC_SAFE_CALL(x)                       \
```
Comment (naoyam) [inline csrc/executor_utils.h:31]: Use the standard assertion in nvFuser makes life slightly easier

File: test/test_gpu3.cpp [hunk 7]
```diff
@@ -1979,13 +1979,8 @@ TEST_F(NVFuserTest, FusionTooLargeSmem_CUDA) {
...
 
-  // First compile gets a compiled kernel
-  fe.compileFusion(&fusion, {t0});
-
-  // Should be throwing because the kernel
-  //  requested absolute device limit
-  // NOLINTNEXTLINE(cppcoreguidelines-avoid-goto,hicpp-avoid-goto)
-  ASSERT_ANY_THROW(fe.runFusion({t0}));
+  //  NOLINTNEXTLINE(cppcoreguidelines-avoid-goto,hicpp-avoid-goto)
+  ASSERT_ANY_THROW(fe.compileFusion(&fusion, {t0}));
```
Comment (naoyam) [inline test/test_gpu3.cpp:1983]: Error check is now done at the compilation time

**Comment by naoyam** [comment]
@mmigdal-nv @jjsjann123 Can you please review this PR?

File: csrc/executor.cpp [hunk 8]
```diff
@@ -366,13 +369,15 @@ void FusionExecutor::compileFusion(
...
 
-  // The driver API call requires an int argument.
-  int max_dynamic_smem = 0;
-  CUDA_SAFE_CALL(cuFuncGetAttribute(
-      &max_dynamic_smem,
-      CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES,
-      compiled_kernel_.function));
-  maybe_available_dynamic_smem_ = max_dynamic_smem;
+  // These should be nullopt at this point, but reset just in case
+  available_dynamic_smem_size_.reset();
```
Comment (jjsjann123) [inline csrc/executor.cpp:373]: IIUC, `available_dynamic_smem_size_` and `static_smem_size_` should always be cleared together when we want to re-compute smem??

nitpick: maybe put them inside a helper function `reset_smem`.

File: csrc/executor.cpp [hunk 9]
```diff
@@ -257,8 +257,9 @@ void FusionExecutor::compileFusion(
   TORCH_INTERNAL_ASSERT(
       options_.device.is_cuda(), "Provided device to CUDA fuser is the CPU.");
   auto properties = at::cuda::getDeviceProperties(options_.device.index());
-  configured_device_smem_ = properties->sharedMemPerBlock;
-  device_smem_limit_ = properties->sharedMemPerBlockOptin;
+  // TODO: These properties should be set as part of the constructor so that it
+  // can be const
+  device_smem_limit_ = static_cast<int64_t>(properties->sharedMemPerBlockOptin);
```
Comment (jjsjann123) [inline csrc/executor.cpp:262]: Naive question: is `sharedMemPerBlockOptin == properties->sharedMemPerBlock` when there's not runtime API overwrite?

**Review by jjsjann123 — APPROVED** [review]
LGTM

**Comment by mmigdal-nv** [comment]
> @mmigdal-nv @jjsjann123 Can you please review this PR?

Thank you for the improvements! Works for me (I can't approve though)

File: csrc/executor.cpp (hunk 9 above)
Comment (naoyam) [reply csrc/executor.cpp:262]: I think so.

File: csrc/executor.cpp (hunk 8 above)
Comment (naoyam) [reply csrc/executor.cpp:373]: Yes. Will add the func

**Pull request merged**